File: Easy/merge_two_sorted_lists/merge_two_sorted_lists.py
from typing import List, Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
    dummy = ListNode()
    curr = dummy

    while list1 and list2:
        if list1.val < list2.val:
            curr.next = list1
            list1 = list1.next
        else:
            curr.next = list2
            list2 = list2.next
        curr = curr.next

    if list1:
        curr.next = list1
    else:
        curr.next = list2

    return dummy.next

#Example usage:
if __name__ == "__main__":
    print("Testing mergeTwoLists function:")
    print("Merging two sorted linked lists: [1 -> 2 -> 4] and [1 -> 3 -> 4]")
    # Creating first sorted linked list: 1 -> 2 -> 4
    list1 = ListNode(1)
    list1.next = ListNode(2)
    list1.next.next = ListNode(4)
    print("First sorted linked list:")
    curr = list1
    while curr:
        print(curr.val, end=" -> ")
        curr = curr.next
    print("None")

    # Creating second sorted linked list: 1 -> 3 -> 4
    list2 = ListNode(1)
    list2.next = ListNode(3)
    list2.next.next = ListNode(4)
    print("Second sorted linked list:")
    curr = list2
    while curr:
        print(curr.val, end=" -> ")
        curr = curr.next
    print("None")

    # Merging the two sorted linked lists
    merged_list = mergeTwoLists(None, list1, list2)

    # Printing the merged linked list
    print("Merged sorted linked list:")
    while merged_list:
        print(merged_list.val, end=" -> ")
        merged_list = merged_list.next
    print("None")  # Indicating the end of the linked list


    #Second test case
    print("\nMerging two sorted linked lists: [] and []")
    # Creating first sorted linked list: []
    list1 = None
    print("First sorted linked list: None")
    # Creating second sorted linked list: []
    list2 = None
    print("Second sorted linked list: None")

    # Merging the two sorted linked lists
    merged_list = mergeTwoLists(None, list1, list2)

    # Printing the merged linked list
    print("Merged sorted linked list:")
    if not merged_list:
        print("None")

    #Third test case
    print("\nMerging two sorted linked lists: [] and [0]")
    # Creating first sorted linked list: []
    list1 = None
    print("First sorted linked list: None")
    # Creating second sorted linked list: [0]
    list2 = ListNode(0)
    print("Second sorted linked list:")
    curr = list2
    while curr:
        print(curr.val, end=" -> ")
        curr = curr.next
    print("None")
    

File: Easy/merge_two_sorted_lists/merge_two_sorted_lists.cpp
#include <iostream>
#include <vector>
using namespace std;

/*
* Definition for singly linked list. */

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
    public:
        ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
if (list1 == nullptr) {
            return list2;
        }

        if (list2 == nullptr) {
            return list1;
        }

        if (list1->val < list2->val) {
            list1->next = mergeTwoLists(list1->next, list2);
            return list1;
        }

        list2->next = mergeTwoLists(list1, list2->next);
        return list2;
    }
};

void printList(ListNode* head) {
    if (head == nullptr) {
        cout << "[]";
        return;
    }

    cout << "[";

    while (head != nullptr) {
        cout << head->val;

        if (head->next != nullptr) {
            cout << ", ";
        }

        head = head->next;
    }

    cout << "]";
}

void deleteList(ListNode* head) {
    while (head != nullptr) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

int main() {
    Solution solution;

    // Test 1: [1,2,4] and [1,3,4]
    ListNode* list1 =
        new ListNode(1, new ListNode(2, new ListNode(4)));

    ListNode* list2 =
        new ListNode(1, new ListNode(3, new ListNode(4)));

    cout << "Test case 1" << endl;
    cout << "List 1: ";
    printList(list1);
    cout << endl;

    cout << "List 2: ";
    printList(list2);
    cout << endl;

    ListNode* mergedList1 = solution.mergeTwoLists(list1, list2);

    cout << "Merged: ";
    printList(mergedList1);
    cout << endl << endl;

    deleteList(mergedList1);


    // Test 2: [] and []
    ListNode* list3 = nullptr;
    ListNode* list4 = nullptr;

    cout << "Test case 2" << endl;
    cout << "List 1: ";
    printList(list3);
    cout << endl;

    cout << "List 2: ";
    printList(list4);
    cout << endl;

    ListNode* mergedList2 = solution.mergeTwoLists(list3, list4);

    cout << "Merged: ";
    printList(mergedList2);
    cout << endl << endl;


    // Test 3: [] and [0]
    ListNode* list5 = nullptr;
    ListNode* list6 = new ListNode(0);

    cout << "Test case 3" << endl;
    cout << "List 1: ";
    printList(list5);
    cout << endl;

    cout << "List 2: ";
    printList(list6);
    cout << endl;

    ListNode* mergedList3 = solution.mergeTwoLists(list5, list6);

    cout << "Merged: ";
    printList(mergedList3);
    cout << endl;

    deleteList(mergedList3);

    return 0;
}